dgraph/cmd/graphql/run.go

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright 2019 Dgraph Labs, Inc. and Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package graphql
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/dgraph-io/dgo"
+	"github.com/dgraph-io/dgo/protos/api"
+	"github.com/dgraph-io/dgraph/x"
+
+	"github.com/vektah/gqlparser/ast"
+	"github.com/vektah/gqlparser/parser"
+	"github.com/vektah/gqlparser/validator"
+	_ "github.com/vektah/gqlparser/validator/rules" // make gql validator init() all rules
+)
+
+type options struct {
+	schemaFile     string
+	alpha          string
+	useCompression bool
+}
+
+type graphqlSchema struct {
+	Type   string    `json:"dgraph.type,omitempty"`
+	Schema string    `json:"dgraph.graphql.schema.schema,omitempty"`
+	Date   time.Time `json:"dgraph.graphql.schema.date,omitempty"`
+}
+
+type graphqlSchemas struct {
+	Schemas []graphqlSchema `json:"graphqlSchemas,omitempty"`
+}
+
+var opt options
+
+var GraphQL x.SubCommand
+
+func init() {
+	GraphQL.Cmd = &cobra.Command{
+		Use:   "graphql",
+		Short: "Run the Dgraph GraphQL tool",
+		Run: func(cmd *cobra.Command, args []string) {
+			defer x.StartProfile(GraphQL.Conf).Stop()
+			run()
+		},
+	}
+
+	GraphQL.EnvPrefix = "DGRAPH_GRAPHQL"
+
+	flags := GraphQL.Cmd.Flags()
+	flags.StringP("alpha", "a", "127.0.0.1:9080",
+		"Comma-separated list of Dgraph alpha gRPC server addresses")
+	flags.BoolP("use_compression", "C", false,
+		"Enable compression on connection to alpha server")
+	flags.StringP("schema", "s", "schema.graphql",
+		"Location of GraphQL schema file")
+
+	cmdInit := &cobra.Command{
+		Use:   "init",
+		Short: "Initializes Dgraph for a GraphQL schema",
+		Args:  cobra.NoArgs,
+		Run: func(cmd *cobra.Command, args []string) {
+			defer x.StartProfile(GraphQL.Conf).Stop()
+			if err := initDgraph(); err != nil {
+				fmt.Printf("Error : %s\n", err)
+				os.Exit(1)
+			}
+		},
+	}
+	cmdInit.Flags().AddFlag(GraphQL.Cmd.Flag("alpha"))
+	cmdInit.Flags().AddFlag(GraphQL.Cmd.Flag("schema"))
+	GraphQL.Cmd.AddCommand(cmdInit)
+
+	// TLS configuration
+	x.RegisterClientTLSFlags(flags)
+}
+
+func run() {
```

Comment:
Again, very long function with too many nested blocks. Should break it down into multiple functions.

Reply:
as above - will get spilt up as more parts implemented